apps/push_originserver_multithread_linux.c

```diff
@@ -0,0 +1,373 @@
+	oc_rep_start_root_object();
+	switch (iface_mask)
+	{
+	case OC_IF_BASELINE:
+	  printf("oic.if.baseline\n");
+		oc_process_baseline_interface(request->resource);
+		/* fall through */
+	case OC_IF_RW:
+	  printf("oic.if.rw\n");
+		oc_rep_set_int(root, brightness, brightness);
+		oc_rep_set_int(root, power, power);
+		break;
+	default:
+		break;
+	}
+	oc_rep_end_root_object();
+	oc_send_response(request, OC_STATUS_OK);
+}
+
+
+static void cb_retrieve_light2(oc_request_t *request, oc_interface_mask_t iface_mask,
+		void *user_data)
+{
+	(void) user_data;
+
+	printf("cb_retrieve_light2() is called\n");
+	oc_rep_start_root_object();
+	switch (iface_mask)
+	{
+	case OC_IF_BASELINE:
+		oc_process_baseline_interface(request->resource);
+		/* fall through */
+	case OC_IF_RW:
+		oc_rep_set_int(root, brightness, brightness2);
+		oc_rep_set_int(root, power, power2);
+		break;
+	default:
+		break;
+	}
+	oc_rep_end_root_object();
+	oc_send_response(request, OC_STATUS_OK);
+}
+
+/* ======== for light #1 ======== */
+static void change_brightness(void)
+{
+	brightness = (brightness+1)%100;
+	/* notify PUSH origin Resource is changed */
+	oc_resource_state_changed(resource_uri, res->device);
+}
+
+static void change_power(void)
+{
+	power += 5;
+	oc_resource_state_changed(resource_uri, res->device);
+}
+
+/* ======== for light #2 ======== */
+static void change_brightness2(void)
+{
+	brightness2 = (brightness2+1)%100;
+	oc_resource_state_changed(resource_uri2, res2->device);
+}
+
+static void change_power2(void)
+{
+	power2 += 5;
+	oc_resource_state_changed(resource_uri2, res2->device);
+}
+
+/* PUSH payload builder */
+void build_light_payload()
+{
+	oc_rep_open_object(root, rep);
+	oc_rep_set_int(rep, power, power);
+	oc_rep_set_int(rep, brightness, brightness);
+	oc_rep_close_object(root, rep);
+}
+
+static void register_resources(void)
+{
+  /* light2 is not pushable */
+	res2 = oc_new_resource(NULL, resource_uri2, 1, 0);
+	oc_resource_bind_resource_type(res2, resource_rt);
+	oc_resource_bind_resource_interface(res2, OC_IF_RW);
+	oc_resource_set_default_interface(res2, OC_IF_RW);
+	oc_resource_set_discoverable(res2, discoverable);
+	oc_resource_set_observable(res2, observable);
+	oc_resource_set_request_handler(res2, OC_GET, cb_retrieve_light2, NULL);
+	oc_add_resource(res2);
+
+	/* light1 is pushable */
+	res = oc_new_resource(NULL, resource_uri, 1, 0);
+	oc_resource_bind_resource_type(res, resource_rt);
+	oc_resource_bind_resource_interface(res, OC_IF_RW);
+	oc_resource_set_default_interface(res, OC_IF_RW);
+	oc_resource_set_discoverable(res, discoverable);
+	oc_resource_set_observable(res, observable);
+	/* make a Resource pushable */
+	oc_resource_set_pushable(res, pushable);
+	oc_resource_set_request_handler(res, OC_GET, cb_retrieve_light1, NULL);
+	/* set PUSH payload builder */
+	res->payload_builder = build_light_payload;
+	oc_add_resource(res);
+}
+
+static void signal_event_loop(void)
+{
+	pthread_mutex_lock(&mutex);
```

Comment:
I've changed this in master - you don't need to hold mutex when calling pthread_cond_signal

Reply:
I'm not clear about this. 
`mutex` is an variable defined in user application. 
how did you modify iotivity-lite codes?

Reply:
Just remove the `pthread_mutex_lock/unlock` lines. See: https://github.com/iotivity/iotivity-lite/commit/4be0e57b32989591bac4e5927cdb878aeebcd4bf .

Reply:
done